DFE-Digital/apply-for-teacher-training
# Add Workload Identity Federation

## Context
  _This is part 1 of the migration to WIF. Part 2 requires us to change the Google ruby library for making requests to BigQuery and that will follow this PR. The documentation references some of the changes that will appear in that PR._ 

      Add Workload Federated Identity

      Add classes and configuration to enable pods to request Service
      Account Impersonation Tokens to make authenticated requests to Google
      Cloud.

      These classes will be used to replace the use of Google JSON
      credentials in order to harden security around our access to BigQuery.

      The change consists of A new namespace WorkloadIdentityFederation which
      includes the configurable module.

      The module is configured in the `workload_identity_federation.rb` initializer.

      The contingent classes require instantiation though the
      UserCredentials class uses a class instance variable to store the
      latest retrieved Google Credentials in memory.

      The implementation uses Faraday as it is considered our default HTTP
      client within the application.

      Included is a testing module `WorkloadIdentityFederationStubs` to stub the requests
      of each contingent class as well as stub the Azure configuration.

## Changes proposed in this pull request

Add library classes to allow for the application to request Google Credentials via Workload Identity Federation.

## Guidance to review

Add library classes to allow for the application to request Google Credentials via Workload Identity Federation.

#### How to verify

This feature will only work on the `worker` or `secondary-worker` pods so you'll need to adjust the Makefile to access those pods.

I tend to edit the Makefile to change the pod that the `make env shell` command accesses.

Change the line that matches this like so:

https://github.com/DFE-Digital/apply-for-teacher-training/blob/main/Makefile#L182

```
-	$(if ${APP_NAME_SUFFIX}, $(eval APP_NAME=apply-${APP_NAME_SUFFIX}-clock-worker), $(eval APP_NAME=apply-${APP_ENV}-clock-worker))
+	$(if ${APP_NAME_SUFFIX}, $(eval APP_NAME=apply-${APP_NAME_SUFFIX}-worker), $(eval APP_NAME=apply-${APP_ENV}-worker))
```


Then use `PR_NUMBER=9689 make review_aks shell` 

You can also inspect the environment variables on the pod with `PR_NUMBER=9689 make review_aks view-app-secrets` to see that `GOOGLE_CLOUD_CREDENTIALS_STATS` is set.

## Things to check

- [ ] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [ ] If this code adds a column to the DB, decide whether it needs to be in analytics yml file or analytics blocklist
- [ ] If it adds a significant user-facing change, is it documented in the [CHANGELOG](CHANGELOG.md)?
- [ ] Inform data insights team due to database changes
- [x] Make sure all information from the Trello card is in here
- [x] Add PR link to Trello card
